### PR TITLE
feat: add networkInterfaces configuration to launchTemplate

### LINKED ISSUE #2026: configure network interfaces
**Tell us about your request**
configure the network interfaces. In particular enable [AssociatePublicIpAddress](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-networkinterface.html#cfn-ec2-launchtemplate-networkinterface-associatepublicipaddress)

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
Our particular requirement is to associate public IPs on some node's interfaces for UDP ingress where there is no LB support avail.
We do not want to enable public IPs an all nodes, that's why we cannot use a public subnet.

**Are you currently working around this issue?**
None, docs even state  "Do not configure network instances in the launch template"

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by myaser**
Fixes aws/karpenter#2026

**Description**
a reduced version of aws/karpenter#3819, 
Add a networkInterfaces struct to the AWSNodeTemplate CRD, and use the parameters to create AWS launch templates.

**How was this change tested?**
unit and integration tests were created; also tested manually on a test cluster

**Does this change impact docs?**
- [x] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by bwagner5** [comment]
I've taken a first pass and then got side tracked by some other issues. I'll pull this down tomorrow and take it for a test drive.

File: pkg/providers/launchtemplate/launchtemplate.go [hunk 1]
```diff
@@ -254,22 +254,34 @@ func (p *Provider) createLaunchTemplate(ctx context.Context, options *amifamily.
...
 		}
+		return nil
 	}
-	return nil
+
+	var networkInterfacesRequest []*ec2.LaunchTemplateInstanceNetworkInterfaceSpecificationRequest
+	for _, networkInterface := range options.NetworkInterfaces {
+		networkInterfacesRequest = append(networkInterfacesRequest, &ec2.LaunchTemplateInstanceNetworkInterfaceSpecificationRequest{
+			AssociatePublicIpAddress: networkInterface.AssociatePublicIPAddress,
+			Groups:                   lo.Map(options.SecurityGroups, func(s v1alpha1.SecurityGroup, _ int) *string { return aws.String(s.ID) }),
```
Comment (bwagner5) [inline pkg/providers/launchtemplate/launchtemplate.go:280]: looks like this is missing the `deviceIndex`

**Comment by bwagner5** [comment]
I've tested this out on my side and verified everything is updating properly with the LaunchTemplate, instance, and drift when changing between params within NetworkInterfaces!

Nice! Glad we can get some functionality in so that we can build on this base!

Remaining items are just to fix the DeviceIndex and Description passing to the NetworkInterfaces and that should also fix the test that is failing.

File: pkg/providers/launchtemplate/launchtemplate.go (hunk 1 above)
Comment (myaser) [reply pkg/providers/launchtemplate/launchtemplate.go:282]: should be fixed now

**Comment by myaser** [comment]
> I've tested this out on my side and verified everything is updating properly with the LaunchTemplate, instance, and drift when changing between params within NetworkInterfaces!
> 
> Nice! Glad we can get some functionality in so that we can build on this base!
> 
> Remaining items are just to fix the DeviceIndex and Description passing to the NetworkInterfaces and that should also fix the test that is failing.

great to hear 🎉 
I have added the missing fields

File: test/suites/integration/networkinterfaces_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+		provisioner := test.Provisioner(test.ProvisionerOptions{ProviderRef: &v1alpha5.MachineTemplateRef{Name: provider.Name}})
+		pod := test.Pod()
+		env.ExpectCreated(pod, provider, provisioner)
+		env.EventuallyExpectHealthy(pod)
+		env.ExpectCreatedNodeCount("==", 1)
+		instance := env.GetInstance(pod.Spec.NodeName)
+		Expect(instance.NetworkInterfaces).To(HaveLen(1))
+		Expect(instance.NetworkInterfaces[0]).ToNot(BeNil())
+		Expect(instance.NetworkInterfaces[0].Attachment).To(HaveField("DeviceIndex", HaveValue(Equal(int64(0)))))
+		Expect(instance.NetworkInterfaces[0].Attachment).To(HaveField("NetworkCardIndex", HaveValue(Equal(int64(0)))))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:55]:
```suggestion
```

File: test/suites/integration/networkinterfaces_test.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:80]:
```suggestion
```

File: test/suites/integration/networkinterfaces_test.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+		Expect(instance.NetworkInterfaces[0].Ipv4Prefixes).To(HaveLen(2))
+		Expect(instance.NetworkInterfaces[0].Ipv6Prefixes).To(HaveLen(2))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:125]:
```suggestion
```

File: test/suites/integration/networkinterfaces_test.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+		pod := test.Pod()
+		env.ExpectCreated(pod, provider, provisioner)
+		env.EventuallyExpectHealthy(pod)
+		env.ExpectCreatedNodeCount("==", 1)
+		instance := env.GetInstance(pod.Spec.NodeName)
+		Expect(instance.NetworkInterfaces).To(HaveLen(2))
+
+		Expect(instance.NetworkInterfaces[0]).ToNot(BeNil())
+		Expect(instance.NetworkInterfaces[0].Attachment).To(HaveField("DeviceIndex", HaveValue(Equal(int64(0)))))
+		Expect(instance.NetworkInterfaces[0].Attachment).To(HaveField("NetworkCardIndex", HaveValue(Equal(int64(0)))))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:160]:
```suggestion
```

File: test/suites/integration/networkinterfaces_test.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+		Expect(instance.NetworkInterfaces[0].Description).To(Equal(desc1))
+		Expect(instance.NetworkInterfaces[0].InterfaceType).To(Equal("interface"))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:162]:
```suggestion
```

File: test/suites/integration/networkinterfaces_test.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+		Expect(instance.NetworkInterfaces[1]).ToNot(BeNil())
+		Expect(instance.NetworkInterfaces[1].Attachment).To(HaveField("DeviceIndex", HaveValue(Equal(int64(1)))))
+		Expect(instance.NetworkInterfaces[1].Attachment).To(HaveField("NetworkCardIndex", HaveValue(Equal(int64(1)))))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:166]:
```suggestion
```

File: test/suites/integration/networkinterfaces_test.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+		Expect(instance.NetworkInterfaces[1].Description).To(Equal(desc2))
+		Expect(instance.NetworkInterfaces[1].InterfaceType).To(Equal("efa"))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:168]:
```suggestion
```

**Review by bwagner5 — COMMENTED** [review]
/karpenter snapshot

**Review by bwagner5 — COMMENTED** [review]
/karpenter snapshot

**Comment by bwagner5** [comment]
Looks like the network interfaces e2e are still failing:

```
2023-08-21T19:16:12.6161277Z [38;5;9m• [FAILED] [51.744 seconds][0m
2023-08-21T19:16:12.6161643Z NetworkInterfaces
2023-08-21T19:16:12.6162973Z [38;5;243m/home/runner/work/karpenter/karpenter/test/suites/integration/networkinterfaces_test.go:32[0m
2023-08-21T19:16:12.6163764Z   [38;5;9m[1m[It] should create a node with more than one NetworkInterface[0m
2023-08-21T19:16:12.6164422Z   [38;5;243m/home/runner/work/karpenter/karpenter/test/suites/integration/networkinterfaces_test.go:124[0m
2023-08-21T19:16:12.6165081Z [38;5;243m------------------------------[0m
2023-08-21T19:16:12.6165298Z 
2023-08-21T19:16:12.6165493Z [38;5;9m[1mSummarizing 2 Failures:[0m
2023-08-21T19:16:12.6166305Z   [38;5;9m[FAIL][0m [0mNetworkInterfaces [38;5;9m[1m[It] should use the specified NetworkInterface[0m
2023-08-21T19:16:12.6167362Z   [38;5;243m/home/runner/work/karpenter/karpenter/test/suites/integration/networkinterfaces_test.go:121[0m
2023-08-21T19:16:12.6168145Z   [38;5;9m[FAIL][0m [0mNetworkInterfaces [38;5;9m[1m[It] should create a node with more than one NetworkInterface[0m
2023-08-21T19:16:12.6168893Z   [38;5;243m/home/runner/work/karpenter/karpenter/test/suites/integration/networkinterfaces_test.go:155[0m
2023-08-21T19:16:12.6169225Z 
2023-08-21T19:16:12.6169482Z [38;5;9m[1mRan 91 of 93 Specs in 5523.602 seconds[0m
2023-08-21T19:16:12.6170324Z [38;5;9m[1mFAIL![0m -- [38;5;10m[1m89 Passed[0m | [38;5;9m[1m2 Failed[0m | [38;5;11m[1m0 Pending[0m | [38;5;14m[1m2 Skipped[0m
```

File: test/suites/integration/networkinterfaces_test.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+		pod := test.Pod()
+		env.ExpectCreated(pod, provider, provisioner)
+		env.EventuallyExpectHealthy(pod)
+		env.ExpectCreatedNodeCount("==", 1)
+		instance := env.GetInstance(pod.Spec.NodeName)
+		Expect(instance.NetworkInterfaces).To(HaveLen(2))
+
+		Expect(instance.NetworkInterfaces[0]).ToNot(BeNil())
+		Expect(instance.NetworkInterfaces[0].Attachment).To(HaveField("DeviceIndex", HaveValue(Equal(int64(0)))))
+		Expect(instance.NetworkInterfaces[0].Description).To(Equal(desc1))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:155]: I think you just need to dereference Description here:

```suggestion
		Expect(*instance.NetworkInterfaces[0].Description).To(Equal(desc1))
```

File: test/suites/integration/networkinterfaces_test.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+		Expect(instance.NetworkInterfaces[1]).ToNot(BeNil())
+		Expect(instance.NetworkInterfaces[1].Attachment).To(HaveField("DeviceIndex", HaveValue(Equal(int64(1)))))
+		Expect(instance.NetworkInterfaces[1].Description).To(Equal(desc2))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:159]: I think you just need to dereference Description here:

```suggestion
		Expect(*instance.NetworkInterfaces[0].Description).To(Equal(desc1))
```

File: test/suites/integration/networkinterfaces_test.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+		})
+		pod := test.Pod()
+		env.ExpectCreated(pod, provider, provisioner)
+		env.EventuallyExpectHealthy(pod)
+		env.ExpectCreatedNodeCount("==", 1)
+		instance := env.GetInstance(pod.Spec.NodeName)
+		Expect(instance.NetworkInterfaces).To(HaveLen(1))
+		Expect(instance.NetworkInterfaces[0]).ToNot(BeNil())
+		Expect(instance.NetworkInterfaces[0].Attachment).To(HaveField("DeviceIndex", HaveValue(Equal(int64(0)))))
+		Expect(instance.NetworkInterfaces[0].Description).To(Equal(desc))
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:121]: I think you just need to dereference Description here:

```suggestion
		Expect(*instance.NetworkInterfaces[0].Description).To(Equal(desc1))
```

**Comment by myaser** [comment]
> Looks like the network interfaces e2e are still failing:

Sorry about that; I did not manage to run the e2e tests myself because of a different environment setup. Previously I had to hack around it and introduce changes to Makefile and test runner, which soon became outdated with upstream changes.

I hope the last commits has fixed it

**Comment by bwagner5** [comment]
> > Looks like the network interfaces e2e are still failing:
> 
> Sorry about that; I did not manage to run the e2e tests myself because of a different environment setup. Previously I had to hack around it and introduce changes to Makefile and test runner, which soon became outdated with upstream changes.
> 
> I hope the last commits have fixed it

No worries! I'll kick it off!

**Review by bwagner5 — COMMENTED** [review]
/karpenter snapshot

File: test/suites/integration/networkinterfaces_test.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+		Expect(instance.NetworkInterfaces).To(HaveLen(1))
+		Expect(instance.NetworkInterfaces[0]).ToNot(BeNil())
+		Expect(instance.NetworkInterfaces[0].Attachment).To(HaveField("DeviceIndex", HaveValue(Equal(int64(0)))))
+		Expect(instance.PublicIpAddress).To(BeNil())
+	})
+	It("should create a default NetworkInterface if none specified, with public IP auto assignment", func() {
+		subnets := env.GetSubnets(map[string]string{"karpenter.sh/discovery": settings.FromContext(env.Context).ClusterName})
+		Expect(len(subnets)).ToNot(Equal(0))
+		allSubnets := lo.Flatten(lo.Values(subnets))
+		ExpectAssociatePublicIPAddressToBe(true, allSubnets...)
```
Comment (bwagner5) [inline test/suites/integration/networkinterfaces_test.go:61]: this needs to be reset after the test runs. Maybe returning a cleanup function that you could defer here. 

```
defer ExpectAssociatedPublicIPAddressToBe(true, allSubnets...)
```

Comment (bwagner5) [reply test/suites/integration/networkinterfaces_test.go]: I went ahead and pushed up this change, we'll see if that fixes the test runs!

**Review by bwagner5 — COMMENTED** [review]
/karpenter snapshot

**Comment by bwagner5** [comment]
Ok, looks like the non-Integ e2e tests failed due to some transient infra issue. But the Integ tests now only have 2 failures, looking at those now:

```
Summarizing 2 Failures:
  [FAIL] NetworkInterfaces [It] should create a default NetworkInterface if none specified, with no public IP auto assignment
  /home/runner/work/karpenter/karpenter/test/suites/integration/networkinterfaces_test.go:50
  [FAIL] Scheduling [It] should provision three nodes for a zonal topology spread
  /home/runner/work/karpenter/karpenter/test/suites/integration/scheduling_test.go:325
```

**Review by bwagner5 — COMMENTED** [review]
/karpenter snapshot

**Review by bwagner5 — COMMENTED** [review]
/karpenter snapshot

**Review by bwagner5 — COMMENTED** [review]
/karpenter snapshot

**Comment by bwagner5** [comment]
I rebased the PR and am going to try to look at the e2e test failures still occurring tomorrow. Hopefully we can get this merged by next week!

**Comment by myaser** [comment]
I have been on vacation, and did not manage to look at the updates here. will do ASAP and fix the e2e tests

**Comment by myaser** [comment]
@bwagner5, I see you already fixed the tests, and it is passing.
now, I merged the latest main and resolved the conflict in `pkg/apis/v1beta1/zz_generated.deepcopy.go`

File: test/suites/integration/networkinterfaces_test.go (hunk 12 above)
Comment (myaser) [reply test/suites/integration/networkinterfaces_test.go]: thanks for taking care of this, it already fixed the test

**Comment by jonathan-innis** [comment]
@myaser What do you think about just launching this feature on the new v1beta1 APIs? That way we don't have to deal with the conversion logic between the v1alpha5 and the v1beta1 APIs when we are adding this feature in.

**Comment by jonathan-innis** [comment]
@myaser Looks like you still need a rebase on this one with the new changes. Happy to help get this PR into a mergable state and get this one in if you need any help.

**Comment by myaser** [comment]
> @myaser Looks like you still need a rebase on this one with the new changes. Happy to help get this PR into a mergable state and get this one in if you need any help.

Thank you for your support 

I resolved the conflict and updated the PR, however, I did not make any changes to the version. But I see that the autogeneration has. However added it to both alpha and beta

**Comment by jmdeal** [comment]
Sorry to have left this dormant for a couple of weeks, I've been able to start taking a look and overall this looks like its in a pretty good state. There are just a few things that we need to take care of and then I think this should be ready to merge. 

1) We've frozen additions to the alpha APIs so we can remove the additions to `awsnodetemplate`. 
2) I'm going to guess this was a rebase error, but it looks like you've added a bunch of tests back to `pkg/providers/launchtemplate/suite_test.go`. These tests have been split into `nodeclass_test` and `nodetemplate_test` respectively and your additional tests should probably live in `nodeclass_test`.
3) Your e2e tests aren't tolerant to the VPC CNI allocating additional ENIs on top of those created by the launch template. 

I'll keep testing on my side but other than those few things, most of which are refactor / rebase tasks, this looks good to me. I know we've left this dormant for a while so I'm more than happy to help with this so we can get this merged.

**Review by jmdeal — COMMENTED** [review]
/karpenter snapshot

**Review by jmdeal — COMMENTED** [review]
/karpenter snapshot

File: pkg/providers/launchtemplate/launchtemplate.go [hunk 13]
```diff
@@ -275,22 +275,36 @@ func (p *Provider) createLaunchTemplate(ctx context.Context, options *amifamily.
...
-func (p *Provider) generateNetworkInterface(options *amifamily.LaunchTemplate) []*ec2.LaunchTemplateInstanceNetworkInterfaceSpecificationRequest {
-	if options.AssociatePublicIPAddress != nil {
-		return []*ec2.LaunchTemplateInstanceNetworkInterfaceSpecificationRequest{
-			{
-				AssociatePublicIpAddress: options.AssociatePublicIPAddress,
-				DeviceIndex:              aws.Int64(0),
-				Groups:                   lo.Map(options.SecurityGroups, func(s v1beta1.SecurityGroup, _ int) *string { return aws.String(s.ID) }),
-			},
+func (p *Provider) generateNetworkInterfaces(options *amifamily.LaunchTemplate) []*ec2.LaunchTemplateInstanceNetworkInterfaceSpecificationRequest {
+	if len(options.NetworkInterfaces) == 0 {
```
Comment (jmdeal) [inline pkg/providers/launchtemplate/launchtemplate.go:284]: I'm thinking we should drop this behavior now, looking at the quoted issue it looks like the desired long term solution was adding these explicit fields. Thoughts @jonathan-innis?

Comment (jonathan-innis) [reply pkg/providers/launchtemplate/launchtemplate.go:284]: I could see us marking this behavior as deprecated and then choosing to drop this behavior entirely at v1. We can add this to the v1 laundry list that is getting tracked in this issue: #4993

File: test/suites/beta/integration/networkinterfaces_test.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+	DescribeTable(
+		"should correctly create NetworkInterfaces",
+		func(interfaces ...*v1beta1.NetworkInterface) {
+			nodeClass.Spec.NetworkInterfaces = interfaces
+			pod := test.Pod()
+			env.ExpectCreated(pod, nodeClass, nodePool)
+			env.EventuallyExpectHealthy(pod)
+			env.ExpectCreatedNodeCount("==", 1)
+			instance := env.GetInstance(pod.Spec.NodeName)
+			for _, interfaceSpec := range interfaces {
```
Comment (jonathan-innis) [inline test/suites/beta/integration/networkinterfaces_test.go:51]: I think that we have a function that grabs the interfaces from the instance? Can we leverage this function rather than grabbing the instance and doing things with checking the fields?

File: test/suites/beta/integration/networkinterfaces_test.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+			if len(interfaces) == 1 && interfaces[0].AssociatePublicIPAddress != nil {
+				if *interfaces[0].AssociatePublicIPAddress {
+					Expect(instance.PublicIpAddress).ToNot(BeNil())
+				} else {
+					Expect(instance.PublicIpAddress).To(BeNil())
+				}
+			}
+		},
+		Entry("when a single interface is specified", &v1beta1.NetworkInterface{
```
Comment (jonathan-innis) [inline test/suites/beta/integration/networkinterfaces_test.go:70]: Each of these tests adds time. I like our bias for code coverage, but I think that we might be able to get away with testing the most complex scenario in the E2E test environment, condensing this down to a single test and then testing a bunch of different edge-case scenarios within the functional/unit testing

File: website/content/en/preview/concepts/nodeclasses.md [hunk 16]
```diff
@@ -29,42 +29,47 @@ metadata:
...
-  
+
   # optional, configures IMDS for the instance
   metadataOptions:
     httpEndpoint: enabled
     httpProtocolIPv6: disabled
     httpPutResponseHopLimit: 2
     httpTokens: required
 
+  # optional, configure network interfaces for the instance
```
Comment (jonathan-innis) [inline website/content/en/preview/concepts/nodeclasses.md:68]: This is missing the `networkInterfaces` header

**Review by jonathan-innis — COMMENTED** [review]
We should consider adding an example to the `examples/v1beta1` directory as well when we add the interfaces to the API

File: pkg/apis/v1beta1/ec2nodeclass.go [hunk 17]
```diff
@@ -312,6 +315,18 @@ type BlockDevice struct {
 	VolumeType *string `json:"volumeType,omitempty"`
 }
 
+type NetworkInterface struct {
```
[the remark this answers is not in the document]
Comment (jmdeal) [reply pkg/apis/v1beta1/ec2nodeclass.go:318]: I'll need to double check but I believe the zero value for `AssociatePublicIPAddress` would not be an acceptable default. If it's not specified in the `EC2NodeClass` we don't want to specify it in the launch template either that way we don't override the subnets behavior.

File: pkg/apis/v1beta1/ec2nodeclass.go [hunk 18]
```diff
@@ -312,6 +315,18 @@ type BlockDevice struct {
...
+
+	// Associates a public IPv4 address with eth0 for a new network interface.
+	AssociatePublicIPAddress *bool `json:"associatePublicIPAddress,omitempty"`
+
+	// A description for the network interface.
+	Description *string `json:"description,omitempty"`
```
[the remark this answers is not in the document]
Comment (jmdeal) [reply pkg/apis/v1beta1/ec2nodeclass.go:324]: It's not really necessary, its used in tests to grab the correct network interface but that can be done with the device index.

File: test/suites/beta/integration/networkinterfaces_test.go (hunk 14 above)
Comment (jmdeal) [reply test/suites/beta/integration/networkinterfaces_test.go:51]: I'll take a look 👍

**Comment by jmdeal** [comment]
Removing `lifecycle/stale`, this PR is still under consideration. There's been some discussion on whether Karpenter is the ideal place for this configuration surface to live, I plan on trying to move on this soon.

**Comment by garvinp-stripe** [comment]
Can you provide some update on the current thought?

**Comment by jmdeal** [comment]
We're hesitant to expand the `EC2NodeClass` to allow for configuring multiple network interfaces if it isn't strictly necessary. There were two separate use cases in the original issue: configuring the `AssignPublicIpAddress` field, and configuring multiple network interfaces for EFA support. `AssignPublicIpAddress` can only be used if you're configuring a single network interface; you can't assign a public IP in this manner to an instance launched with multiple network interfaces. As for EFA, we currently have a [PR](https://github.com/aws/karpenter/pull/5068) out which enables launching EFA instances without the need for explicit network interface configuration. The path forward here will likely be scoping this PR back to only include an `AssociatePublicIpAddress` field on the `EC2NodeClass` rather than a `NetworkInterfaces` field.

**Comment by garvinp-stripe** [comment]
We might be the odd man out however we do leverage attaching multiple network interfaces to our instances outside of Public IP and EFA, we use a custom CNI that expects a secondary ENI. This PR solves this however we could get around this issue. 

One thing I am curious about Karpenter's position is that the AWS provider variant of Karpenter essentially replaces ASGs and Launch Templates in AWS for us however there is a feature disparity between ASGs and Karpenter node management. Its true we may not need to match the functionalities however in my opinion the bias should be towards closing the disparity rather than away. This isn't specific to network interface support but rather as a whole how does/ should Karpenter think about what features it should be keeping from the components it is replacing vs dropping.

**Comment by jmdeal** [comment]
@garvinp-stripe are you able to elaborate about your use-case, preferably in an issue where it's easier to track? We're certainly not against bringing Karpenter closer to feature parity with launch templates, but we want to drive those decisions by user need rather than adding features because they are in launch templates. In this case, the problems outlined in the original issue either didn't need the full scope or were solved through other mechanisms. If you have a use-case for this we're more than happy to continue the discussion.

**Comment by myaser** [comment]
> The path forward here will likely be scoping this PR back to only include an `AssignPublicIpAddress` field on the `EC2NodeClass` rather than a `NetworkInterfaces` field.

@jmdeal 

ok, I can work on that. but I will do a separate PR.

**Comment by jmdeal** [comment]
Sounds good! I'll go ahead and close this PR out then.
